scaletest/dashboard/chromedp.go

```diff
@@ -0,0 +1,181 @@
+		}
+		if !found {
+			continue
+		}
+		matchedLabel = l
+		break
+	}
+	if !found {
+		return "", nil, xerrors.Errorf("no matches found")
+	}
+
+	return "click_" + matchedLabel, func(ctx context.Context) error {
+		if err := clickAndWait(ctx, matched); err != nil {
+			return xerrors.Errorf("click %q: %w", matched, err)
+		}
+		return nil
+	}, nil
+}
+
+// randMatch returns a random match for the given selector.
+// The returned selector is the full XPath of the matched node.
+// If no matches are found, an error is returned.
+// If multiple matches are found, one is chosen at random.
+func randMatch(ctx context.Context, s Selector) (Selector, bool, error) {
+	var nodes []*cdp.Node
+	err := chromedp.Run(ctx, chromedp.Nodes(s, &nodes, chromedp.NodeVisible, chromedp.AtLeast(0)))
+	if err != nil {
+		return "", false, xerrors.Errorf("get nodes for selector %q: %w", s, err)
+	}
+	if len(nodes) == 0 {
+		return "", false, nil
+	}
+	n := pick(nodes)
+	return Selector(n.FullXPath()), true, nil
+}
+
+// clickAndWait clicks the given selector and waits for the page to finish loading.
+// The page is considered loaded when the network event "LoadingFinished" is received.
+func clickAndWait(ctx context.Context, s Selector) error {
+	return chromedp.Run(ctx, chromedp.Tasks{
+		chromedp.Click(s, chromedp.NodeVisible),
+		chromedp.ActionFunc(func(ctx context.Context) error {
+			return waitForEvent(ctx, func(e interface{}) bool {
+				if _, ok := e.(*network.EventLoadingFinished); ok {
+					return true
+				}
+				return false
+			})
+		}),
+	})
+}
+
+// initChromeDPCtx initializes a chromedp context with the given session token cookie
+//
+//nolint:revive // yes, headless is a control flag
+func initChromeDPCtx(ctx context.Context, u *url.URL, sessionToken string, headless bool) (context.Context, context.CancelFunc, error) {
+	dir, err := os.MkdirTemp("", "scaletest-dashboard")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	allocOpts := append(chromedp.DefaultExecAllocatorOptions[:],
+		chromedp.UserDataDir(dir),
+		chromedp.DisableGPU,
+	)
+
+	if !headless { // headless is the default
+		allocOpts = append(allocOpts, chromedp.Flag("headless", false))
+	}
+
+	allocCtx, allocCtxCancel := chromedp.NewExecAllocator(ctx, allocOpts...)
+	cdpCtx, cdpCancel := chromedp.NewContext(allocCtx)
+	cancelFunc := func() {
+		cdpCancel()
+		allocCtxCancel()
+		_ = os.RemoveAll(dir)
+	}
+
+	// set cookies
+	if err := setSessionTokenCookie(cdpCtx, sessionToken, u.Host); err != nil {
+		cancelFunc()
+		return nil, nil, xerrors.Errorf("set session token cookie: %w", err)
+	}
+
+	// visit main page
+	if err := visitMainPage(cdpCtx, u); err != nil {
+		cancelFunc()
+		return nil, nil, xerrors.Errorf("visit main page: %w", err)
+	}
+
+	return cdpCtx, cancelFunc, nil
+}
+
+func setSessionTokenCookie(ctx context.Context, token, domain string) error {
+	exp := cdp.TimeSinceEpoch(time.Now().Add(30 * 24 * time.Hour))
```

Comment:
1 day max? unless we need them to be valid longer

Reply:
👍